File: 46. Permutations.js
var permute = function(nums) {
    var result = [];
    var visited = [];
    for(var i = 0; i < nums.length; ++i) visited.push(false);
    dfs(nums, [], result, nums.length, visited);
    return result;
};
function dfs(nums, line, result, n, visited) {
    if(!n) {
        result.push(line.slice());
        return;
    }
    for(var i = 0; i < nums.length; ++i) {
        if(visited[i]) continue;
        visited[i] = true;
        line.push(nums[i]);
        dfs(nums, line, result, n-1, visited);
        line.pop();
        visited[i] = false;
    }
}

File: Leetcode/Three/46. Permutations.java
public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        boolean[] isUsed = new boolean[nums.length];
        dfs(result, nums, isUsed, new ArrayList<Integer>());
        return result;
    }
    private void dfs(List<List<Integer>> result, int[] nums, boolean[] isUsed, List<Integer> line) {
        if(line.size() == nums.length) {
            result.add(new ArrayList<Integer>(line));
            return;
        }
        for(int i = 0; i <nums.length; ++i) {
            if(isUsed[i]) continue;
            isUsed[i] = true;
            line.add(nums[i]);
            dfs(result, nums, isUsed,line);
            isUsed[i] = false;
            line.remove(line.size()-1);
        }
    }
}
